CAMEL-15510: Add camel-vertx-kafka component

omarsmak opened the pull request:
This is pretty huge PR but I can summarise the changes here , this PR will add the component of camel-vertx-kafka to Camel, it includes the following:

- A maven plugin that will generate the component configuration from Kafka Consumer/Producer native configurations similar to camel-debezium. Later, I am planning as well to use this plugin to generate the configurations for camel-kafka. This plugin is highly customizable with many options that allows to be adopted it easily to camel-kafka. 
- The component itself, based on Vert.x client for Kafka that supports full async operations, coupled with Mono and Flux just to minimize the callbacks (not necessary) though. In vertx 4.0 (still beta) it uses Future instead of callback, however the work should't be large, with Flux/Mono could be easily wrapped. 
- The component doesnt yet support manual commit offsets, storing offsets to be committed but it can be added later.
- To ensure the consistently functionality, I have used the integration tests and unit tests of camel-kafka to test this component (with small modifications specific to this component).

jamesnetherton commented on components/camel-vertx-kafka/camel-vertx-kafka-component/src/main/java/org/apache/camel/component/vertx/kafka/VertxKafkaComponent.java:
```diff
@@ -0,0 +1,141 @@
+        if (vertx != null) {
+            vertx.close();
+        }
+        vertx = null;
```
Might be better to do like [`VertxComponent.doStop()`](https://github.com/apache/camel/blob/master/components/camel-vertx/src/main/java/org/apache/camel/component/vertx/VertxComponent.java#L201-L209) and handle the case where the Vert.x instance was provided by the user. In that scenario you probably don't want to clean it up as there may be other things that depend on it.

davsclaus commented on components/camel-vertx-kafka/camel-vertx-kafka-component/src/main/java/org/apache/camel/component/vertx/kafka/VertxKafkaConsumer.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    private void onErrorListener(final Throwable error) {
+        final Exchange exchange = getEndpoint().createExchange();
```
Looks like here you need to set the exception on the created exchange as otherwise its null

omarsmak replied:
Ah good note. Didnt realize that it will be always null

davsclaus commented on components/camel-vertx-kafka/camel-vertx-kafka-component/src/main/java/org/apache/camel/component/vertx/kafka/VertxKafkaConsumer.java:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    private void onEventListener(final KafkaConsumerRecord<Object, Object> record) {
+        final Exchange exchange = getEndpoint().createVertxKafkaExchange(record);
```
the name are createExchange on endpoints in the other components

davsclaus commented on components/camel-vertx-kafka/camel-vertx-kafka-component/src/main/java/org/apache/camel/component/vertx/kafka/VertxKafkaHeaderFilterStrategy.java:
```diff
@@ -0,0 +1,43 @@
+
+import org.apache.camel.support.DefaultHeaderFilterStrategy;
+
+public class VertxKafkaHeaderFilterStrategy extends DefaultHeaderFilterStrategy {
```
If you use the header filter strategy, then the endpoint implementation should implement an interface or extend another base class that has support for header filtering, where you can plugin a custom strategy.

omarsmak commented:
@jamesnetherton @davsclaus I have addressed your comments, please take a look. In regards to the filter strategy, I have added the support for the custom header filtering in the configuration where you can plug-in the custom header filter in the configurations.
